modules/fledge.js

```diff
@@ -0,0 +1,192 @@
+/**
+ * Fledge modules is responsible for registering fledged auction configs and
+ * run the fledge auction.
+ */
+import { auctionManager } from '../src/auctionManager.js';
+import { config } from '../src/config.js';
+import { getHook } from '../src/hook.js';
+import { createInvisibleIframe, insertElement, getGptSlotForAdUnitCode, logInfo, logWarn, logError } from '../src/utils.js';
+
+const MODULE = 'fledge'
+
+// should be an array of {bid.adUnitCode, seller, auctionConfig}
+// where there is only one entry for each {bid.adUnitCode, seller} tuple
+// TODO do we want to keep track of entries per auctionId or just the last one ?
+let _componentAuctions = [];
+
+config.getConfig('fledge', config => init(config.fledge));
+
+/**
+  * Module init.
+  */
+export function init(fledgeCfg) {
+  logInfo(MODULE, `init`, fledgeCfg);
+  if (fledgeCfg.enabled !== false) {
+    if (fledgeCfg.useGpt) {
+      getHook('addComponentAuction').before(addComponentAuctionToGptHook);
+    } else {
+      getHook('addComponentAuction').before(addComponentAuctionHook);
+      getHook('renderAd').before(renderAdHook);
+      getHook('secureRenderRequest').before(secureRenderRequestHook);
+    }
+  }
+}
+
+export function addComponentAuctionToGptHook(next, bidRequest, componentAuctionConfig) {
+  const seller = componentAuctionConfig.seller;
+  const adUnitCode = bidRequest.adUnitCode;
+  const gptSlot = getGptSlotForAdUnitCode(adUnitCode);
+  logWarn(MODULE, 'gptSlotInfo', gptSlot);
+  if (gptSlot && gptSlot.setConfig) {
+    delete componentAuctionConfig.bidId;
+    gptSlot.setConfig({
+      componentAuction: [{
+        configKey: seller,
+        auctionConfig: componentAuctionConfig
+      }]
+    });
+    logInfo(MODULE, `register component auction config in GPT for: ${adUnitCode} x ${seller}: ${gptSlot.getAdUnitPath()}`, componentAuctionConfig);
+  } else {
+    logWarn(MODULE, `unable to register component auction config in GPT for: ${adUnitCode} x ${seller}.`);
+  }
+
+  next(bidRequest, componentAuctionConfig);
+}
+
+export function addComponentAuctionHook(next, bidRequest, fledgeAuctionConfig) {
+  if (!bidRequest || !fledgeAuctionConfig) { return; }
+
+  const seller = fledgeAuctionConfig.seller;
+  const adUnitCode = bidRequest.adUnitCode;
+
+  // TODO : remove duplicate {adUnitCode, seller} is already registered
+  logInfo(MODULE, `adding auction config for ${adUnitCode} x ${seller}`);
+  _componentAuctions.push({adUnitCode, seller, fledgeAuctionConfig});
+
+  next(bidRequest, fledgeAuctionConfig);
+}
+
+export function getAuctionConfig(bid) {
+  const targetAdUnitCode = bid.adUnitCode;
+  const componentAuctionsForAdUnitCode = _componentAuctions.filter((componentAuction) => {
```

Comment:
you can run an auction with multiple adUnits that have the same code, you can also run multiple auctions on the same adUnits. Does this approach work for those cases? 

bids have a `transactionId` that might work better as a filter. They are unique for each auction-adUnit pair, and you should have their association with the fledge configs through their `bidId`.

Reply:
`adUnitCode` is needed to find the GPT slot  associated with the adUnit. With GAM, which may well be most of the use cases, the fledge auction config is set on the GPT slot and GPT triggers the auction.

When GAM isn't in use, the configs could be organized a different way, true, but this depends on what the buyers expect to bid on and those bidding functions would then need to take into account the whole environment they bid in. 

Would you happen to know the use case for sharing `adUnitCode` across different adunits ? Is this a common case ?

Reply:
[Twin ad unit codes](https://docs.prebid.org/dev-docs/adunit-reference.html#twin-adunit-codes). 

Even when using GAM, it looks easier to me to filter by transactionId. If you don't you'll have to filter by adUnitCode and auctionId, and that does not cover the repeated adUnitCode case.
l